subsys/net/lib/sockets/sockets.c

```diff
@@ -209,45 +210,100 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 	return POINTER_TO_INT(ctx);
 }
 
+static struct net_pkt* get_tx_pkt(struct net_context *ctx,
+				  const void *buf, size_t len,
+				  s32_t timeout)
+{
+	struct net_pkt *pkt = NULL;
+	size_t max_len = net_if_get_mtu(net_context_get_iface(ctx));
+
+	pkt = net_pkt_get_tx(ctx, timeout);
+	if (!pkt) {
+		return NULL;
+	}
+
+	/* Make sure we don't send more data in one packet than
+	 * MTU allows. Optimize for number of branches in the code.
+	 */
+	max_len -= NET_IPV4TCPH_LEN;
+	if (net_context_get_family(ctx) != AF_INET) {
+		max_len -= NET_IPV6TCPH_LEN - NET_IPV4TCPH_LEN;
+	}
+
+	if (len > max_len) {
+		len = max_len;
+	}
+
+	len = net_pkt_append(pkt, len, buf, timeout);
+	if (!len) {
+		net_pkt_unref(pkt);
+		return NULL;
+	}
+
+	return pkt;
+}
+
 ssize_t zsock_send(int sock, const void *buf, size_t len, int flags)
 {
 	ARG_UNUSED(flags);
 	int err;
 	struct net_pkt *send_pkt;
 	s32_t timeout = K_FOREVER;
 	struct net_context *ctx = INT_TO_POINTER(sock);
-	size_t max_len = net_if_get_mtu(net_context_get_iface(ctx));
 
 	if (sock_is_nonblock(ctx)) {
 		timeout = K_NO_WAIT;
 	}
 
-	send_pkt = net_pkt_get_tx(ctx, timeout);
+	send_pkt = get_tx_pkt(ctx, buf, len, timeout);
 	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	/* Make sure we don't send more data in one packet than
-	 * MTU allows. Optimize for number of branches in the code.
-	 */
-	max_len -= NET_IPV4TCPH_LEN;
-	if (net_context_get_family(ctx) != AF_INET) {
-		max_len -= NET_IPV6TCPH_LEN - NET_IPV4TCPH_LEN;
+	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	if (err < 0) {
+		net_pkt_unref(send_pkt);
+		errno = -err;
+		return -1;
 	}
 
-	if (len > max_len) {
-		len = max_len;
+	return len;
+}
+
+ssize_t zsock_sendto(int sock, const void *buf, size_t len, int flags,
+		     const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	ARG_UNUSED(flags);
+	int err;
+	struct net_pkt *send_pkt;
+	s32_t timeout = K_FOREVER;
+	struct net_context *ctx = INT_TO_POINTER(sock);
+
+	if (sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
 	}
 
-	len = net_pkt_append(send_pkt, len, buf, timeout);
-	if (!len) {
-		net_pkt_unref(send_pkt);
+	send_pkt = get_tx_pkt(ctx, buf, len, timeout);
+	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	/* auto bind if necessary */
+	if (!net_context_is_bound(ctx)) {
+		struct sockaddr src_addr;
+		memset(&src_addr, 0, sizeof(src_addr));
```

Comment:
lines break are cheap and help for code readability, add empty lines after variable declaration.